### PR TITLE
API request and data presentation

### PR DESCRIPTION
**Pull request by LilianaFaustinoBloco**

### CODE REVIEW

File: template/androidApp/src/main/java/io/bloco/template/features/details/DetailsScreen.kt [hunk 1]
```diff
@@ -1,9 +1,42 @@
...
+import io.bloco.template.component.Toast
 
 @Composable
-fun DetailsScreen(detailsViewModel: DetailsViewModel, id: String?) {
-    Text("Received: $id")
+fun DetailsScreen(detailsViewModel: DetailsViewModel, id: String) {
+    val bookListUpdateState by detailsViewModel.bookDetailsUpdateState.collectAsState()
+    val bookDetails by detailsViewModel.bookDetails.collectAsState()
+
+    //TODO: Talvez a solução seja https://medium.com/scalereal/providing-assistedinject-supported-viewmodel-for-composable-using-hilt-ae973632e29a
```
Comment (Filmaluco) [inline template/androidApp/src/main/java/io/bloco/template/features/details/DetailsScreen.kt:22]: Yes, seems a better approach :) let's go for it

File: template/androidApp/src/main/java/io/bloco/template/features/details/DetailsViewModel.kt [hunk 2]
```diff
@@ -1,11 +1,54 @@
...
 import javax.inject.Inject
 
 @HiltViewModel
 class DetailsViewModel @Inject constructor(
+    getBook: GetBook
+) : ViewModel() {
+    private val events = PublishFlow<Event>()
 
-): ViewModel() {
+    private val _bookDetails = MutableStateFlow(BookDetails(""))
```
Comment (Filmaluco) [inline template/androidApp/src/main/java/io/bloco/template/features/details/DetailsViewModel.kt:22]: we can start using 1.7 here I think

think we can use internals here 
imagem

File: template/androidApp/src/main/java/io/bloco/template/features/details/DetailsViewModel.kt [hunk 3]
```diff
@@ -1,11 +1,54 @@
...
+            getBook(event.id).onSuccess {
+                _bookDetails.emit(it)
+                _bookDetailsUpdateState.value = APIRequestState.Success
+            }.onFailure {
+                _bookDetailsUpdateState.value = APIRequestState.ErrorFromAPI
+            }
+        }.launchIn(viewModelScope)
+    }
+
+    fun sendBookId(id: String) {
```
Comment (Filmaluco) [inline template/androidApp/src/main/java/io/bloco/template/features/details/DetailsViewModel.kt:41]: acho que isto com a abordagem de usar @assistendInject como explicado no artigo acima fica mais clean.

File: template/androidApp/src/main/java/io/bloco/template/features/list/ListScreen.kt [hunk 4]
```diff
@@ -1,20 +1,61 @@
...
-    /*TemplateButton(onClick = {
-        openDetailsClicked("1")
-    }) {
-        Text(text = "Open Details")
+    val bookListUpdateState by listViewModel.bookListUpdateState.collectAsState()
+    val books = listViewModel.books.collectAsState()
+
+    when (bookListUpdateState) {
+        ListViewModel.BookListUpdateState.ErrorFromAPI -> Toast(R.string.api_error)
+        ListViewModel.BookListUpdateState.LoadingFromAPI -> Unit
```
Comment (Filmaluco) [inline template/androidApp/src/main/java/io/bloco/template/features/list/ListScreen.kt:32]: Not even a circular progress bar?

File: template/androidApp/src/main/java/io/bloco/template/features/list/ListScreen.kt [hunk 5]
```diff
@@ -1,20 +1,61 @@
...
 fun ListScreen(listViewModel: ListViewModel, openDetailsClicked: (String) -> Unit) {
-    /*TemplateButton(onClick = {
-        openDetailsClicked("1")
-    }) {
-        Text(text = "Open Details")
+    val bookListUpdateState by listViewModel.bookListUpdateState.collectAsState()
+    val books = listViewModel.books.collectAsState()
+
+    when (bookListUpdateState) {
+        ListViewModel.BookListUpdateState.ErrorFromAPI -> Toast(R.string.api_error)
```
Comment (Filmaluco) [inline template/androidApp/src/main/java/io/bloco/template/features/list/ListScreen.kt:31]: toast goes away, maybe it would be better a screen with a message, check wifi something like that. 
It would also be great make a retry action

File: template/androidApp/src/main/java/io/bloco/template/navigation/NavHost.kt [hunk 6]
```diff
@@ -20,7 +20,7 @@ fun TemplateNaveHost(
             })
         }
         composable("/detail{id}") { backStackEntry ->
```
Comment (Filmaluco) [inline template/androidApp/src/main/java/io/bloco/template/navigation/NavHost.kt:22]: magic strings 😠

File: template/androidApp/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -1,3 +1,6 @@
 <resources>
     <string name="app_name">template</string>
+    <string name="api_error">Error from API.</string>
+    <string name="book_list">Book list</string>
+    <string name="book_title">Book title:</string>
```
Comment (Filmaluco) [inline template/androidApp/src/main/res/values/strings.xml:5]: if we assume there's something after, a formated string would be better

File: template/core/data/src/main/java/io/bloco/core/data/models/BookDetailsDto.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package io.bloco.core.data.models
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class BookDetailsDto(
+    @SerialName("title")
```
Comment (Filmaluco) [inline template/core/data/src/main/java/io/bloco/core/data/models/BookDetailsDto.kt:8]: I am guessing u did this for consistence, but
**TIP**: if the name is the same as the variable you don't need the annotation :P

File: template/core/data/src/main/java/io/bloco/core/data/repositories/BookRepository.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import io.bloco.core.data.models.BookDetailsDto
+import io.bloco.core.data.models.BookDto
+import io.bloco.core.data.network.OpenLibraryService
+import javax.inject.Inject
+
+class BookRepository
+@Inject constructor(
+    private val openLibraryService: OpenLibraryService
+) {
+    suspend fun getBooks(): Result<List<BookDto>> =
```
Comment (Filmaluco) [inline template/core/data/src/main/java/io/bloco/core/data/repositories/BookRepository.kt:12]: Here something we sometimes do is create a CollectionOfBooksDto. This allows for us to use pagination later one with lesser refactors. But I guess it might be overkill... not sure...

File: template/core/domain/src/main/java/io/bloco/core/domain/GetBook.kt [hunk 10]
```diff
@@ -1,4 +1,18 @@
...
+import io.bloco.core.domain.models.toModel
+import javax.inject.Inject
+
+class GetBook @Inject constructor(
+    private val bookRepository: BookRepository
+
+) {
+    suspend operator fun invoke(id: String): Result<BookDetails> {
+        return bookRepository.getBookDetails(id).map { bookDetails ->
+            println("Title: " + bookDetails.title)
```
Comment (Filmaluco) [inline template/core/domain/src/main/java/io/bloco/core/domain/GetBook.kt:14]: println? WHAT IS THIS 😆

**Review by Filmaluco — APPROVED** [review]
Packages still need an update, but nothing blocking :)

File: template/androidApp/src/main/java/io/bloco/template/features/details/DetailsViewModel.kt (hunk 2 above)
Comment (Filmaluco) [reply template/androidApp/src/main/java/io/bloco/template/features/details/DetailsViewModel.kt:22]: It was a good idea, just tested it. 
Still experimental and needs K2 compiler, currently it doesn't seem to be compatible with the Android plugin

File: template/androidApp/src/androidTest/java/io/bloco/template/ExampleInstrumentedTest.kt [hunk 11]
```diff
@@ -15,10 +19,21 @@ import org.junit.Assert.*
  */
 @RunWith(AndroidJUnit4::class)
 class ExampleInstrumentedTest {
+
+    @get:Rule
+    val composeTestRule = createComposeRule()
+
+    init {
+        io.bloco.core.commons.RootLogger.setup({ _: Level, _: String, message: String, _: Throwable? ->
```
Comment (sdsantos) [inline template/androidApp/src/androidTest/java/io/bloco/template/ExampleInstrumentedTest.kt:27]: Could we import `io.bloco.core.commons.RootLogger` instead?

Comment (sdsantos) [reply template/androidApp/src/androidTest/java/io/bloco/template/ExampleInstrumentedTest.kt:27]: Also, ideally, by default the RootLogger comes with this logger, and it is swapped out if any other logger is added.

File: template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    fun logError(throwable: Throwable, message: String)
+}
+
+fun Any.logDebug(message: () -> String) {
+    if (allowDebugLog) {
+        loggers.forEach {
+            it.logDebug(message())
+        }
+        logger.log(Level.FINE, message())
+    }
```
Comment (sdsantos) [inline template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt:76]: Isn't this double logging? And why do we need this flag, if the TemplateLogger implementation can already do this for us?

File: template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+fun Any.logWarning(message: () -> String) {
+    loggers.forEach { it.logWarning(message()) }
+    logger.log(Level.WARNING, message())
+}
```
Comment (sdsantos) [inline template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt:82]: Warning can also have throwables.

File: template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        override fun flush() {}
+        override fun close() {}
+    }
+
+    for (h in RootLogger.handlers) {
+        removeHandler(h)
+    }
+    addHandler(handler)
+    allowDebugLog = allowDebug
+    level = Level.FINE
```
Comment (sdsantos) [inline template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt:56]: Why is this needed?

File: template/androidApp/src/main/java/io/bloco/template/App.kt [hunk 15]
```diff
@@ -1,9 +1,59 @@
...
+            val androidLogLevel = when (level.intValue()) {
+                Level.SEVERE.intValue() -> Log.ERROR
+                Level.WARNING.intValue() -> Log.WARN
+                Level.INFO.intValue() -> Log.INFO
+                else -> Log.DEBUG
+            }
+
+            error?.let {
+                Log.e(tag, message, error)
+            } ?: Log.println(androidLogLevel, tag, message)
```
Comment (sdsantos) [inline template/androidApp/src/main/java/io/bloco/template/App.kt:55]: Warning logs can have errors too.

File: template/androidApp/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -1,3 +1,12 @@
...
     <string name="app_name">template</string>
+
+    <!-- Generic -->
+    <string name="api_error">Error from API.</string>
+
+    <!-- List Screen -->
+    <string name="book_list">Book list</string>
+
+    <!-- Details Screen -->
+    <string name="book_title">Book title: %s</string>
```
Comment (sdsantos) [inline template/androidApp/src/main/res/values/strings.xml:11]: Avoid these strings if possible. Things might look worse, but it's fewer things to delete later.

File: template/androidApp/src/test/java/io/bloco/template/ExampleUnitTest.kt [hunk 17]
```diff
@@ -1,17 +1,28 @@
...
 class ExampleUnitTest {
+    init {
+        io.bloco.core.commons.RootLogger.setup({ _: Level, _: String, message: String, _: Throwable? ->
+            println(message)
+        }, true)
+    }
+
     @Test
     fun addition_isCorrect() {
+        logDebug { "something" }
```
Comment (sdsantos) [inline template/androidApp/src/test/java/io/bloco/template/ExampleUnitTest.kt:25]: I get that this is a test, but then write an actual test to check the RootLogger is working. But it's better to not have this at all, and have an actual unit test.

File: template/build-logic/convention/src/main/kotlin/KotlinFeatureConventionPlugin.kt [hunk 18]
```diff
@@ -1,14 +1,54 @@
...
+            dependencies {
+                add("implementation", libs.findLibrary("coroutines").get())
+            }
         }
     }
 }
 
+
+
+
```
Comment (sdsantos) [inline template/build-logic/convention/src/main/kotlin/KotlinFeatureConventionPlugin.kt:54]: [ this space shouldn't have been left intentionally blank ]

File: template/build.gradle.kts [hunk 19]
```diff
@@ -1,13 +1,49 @@
...
+                }
+            }
+        }
+    }
+
+    // optional parameters
+    checkForGradleUpdate = true
+    outputFormatter = "html"
+    outputDir = "build/dependencyUpdates"
+    reportfileName = "report"
```
Comment (sdsantos) [inline template/build.gradle.kts:48]: It's hard to tell what is this for.

File: template/core/commons/src/main/java/io/bloco/core/commons/ResultUtils.kt [hunk 20]
```diff
@@ -0,0 +1,38 @@
+package io.bloco.core.commons
```
Comment (sdsantos) [inline template/core/commons/src/main/java/io/bloco/core/commons/ResultUtils.kt:1]: I get that these Utils are useful, but could we do without them in the Template? It's not a sure thing that we will use the Kotlin Result in all our projects...

File: template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryHttpClient.kt [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+                })
+            }
+
+            install(Logging) {
+                logger = object : Logger {
+                    override fun log(message: String) {
+                        logDebug { "Response: $message" }
+                    }
+                }
+                level = LogLevel.ALL
```
Comment (sdsantos) [inline template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryHttpClient.kt:43]: This is not a good default for a release, we should lower the level for release builds.

File: template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryHttpClient.kt [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+import io.ktor.http.ContentType
+import io.ktor.http.HttpHeaders
+import io.ktor.http.URLProtocol
+import io.ktor.serialization.kotlinx.json.json
+import kotlinx.serialization.json.Json
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class OpenLibraryHttpClient @Inject constructor() {
```
Comment (sdsantos) [inline template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryHttpClient.kt:22]: Maybe we should call this a more generic name, and extract the HOST into a variable. Those are definitely things that will be changed on a new project.

File: template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt [hunk 23]
```diff
@@ -1,34 +1,40 @@
...
-    suspend fun getBooks(): Result<BookRecords> = try {
-        Result.success(
-            httpClient.get {
-                url(path = "/")
-            }.body()
-        )
-    } catch (e: Exception) {
-        Result.failure(e)
+
+    suspend fun getBooks(): Result<BookRecords> = withContext(Dispatchers.IO) {
```
Comment (sdsantos) [inline template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt:17]: I believe ktor already handles the network calls in a background dispatcher, so we don't need to change dispatchers here. Also, never hardcode a scope/context/dispatcher, it always should be swappable, or at least exposed.

File: template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt [hunk 24]
```diff
@@ -1,34 +1,40 @@
...
+        return@withContext try {
+            Result.success(
+                httpClient().get {
+                    url(path = "/search.json?q=android&limit=10")
+                }.body()
+            )
+        } catch (e: Exception) {
```
Comment (sdsantos) [inline template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt:24]: Is there a more specific exception?

File: template/core/domain/src/main/java/io/bloco/core/domain/GetBook.kt [hunk 25]
```diff
@@ -1,4 +1,19 @@
...
\ No newline at end of file
+import io.bloco.core.commons.logDebug
+import io.bloco.core.data.repositories.BookRepository
+import io.bloco.core.domain.models.BookDetails
+import io.bloco.core.domain.models.toModel
+import javax.inject.Inject
+
+class GetBook @Inject constructor(
+    private val bookRepository: BookRepository
+
```
Comment (sdsantos) [inline template/core/domain/src/main/java/io/bloco/core/domain/GetBook.kt:11]: Extra line.

File: template/core/domain/src/main/java/io/bloco/core/domain/GetBook.kt [hunk 26]
```diff
@@ -1,4 +1,19 @@
...
+) {
+    suspend operator fun invoke(id: String): Result<BookDetails> {
+        return bookRepository.getBookDetails(id).map { bookDetails ->
+            logDebug { ("Title: " + bookDetails.title) }
```
Comment (sdsantos) [inline template/core/domain/src/main/java/io/bloco/core/domain/GetBook.kt:15]: Parenthesis not needed.

File: template/core/domain/src/main/java/io/bloco/core/domain/models/Book.kt [hunk 27]
```diff
@@ -0,0 +1,15 @@
...
+    val key: String,
+    val title: String
+)
+
+fun BookDto.toModel(): Book = Book(
+    key = key.removePrefix(Key_Prefix),
+    title = title
+)
+
+private const val Key_Prefix = "/works/"
```
Comment (sdsantos) [inline template/core/domain/src/main/java/io/bloco/core/domain/models/Book.kt:15]: Why do we need to remove the prefix?

**Review by sdsantos — APPROVED** [review]

File: template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt (hunk 23 above)
Comment (Filmaluco) [reply template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt:17]: strict mode was crashing, hence the change. 
But yes I shall make it a argument

File: template/core/domain/src/main/java/io/bloco/core/domain/models/Book.kt (hunk 27 above)
Comment (Filmaluco) [reply template/core/domain/src/main/java/io/bloco/core/domain/models/Book.kt:15]: @LilianaFaustinoBloco

File: template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt (hunk 23 above)
Comment (sdsantos) [reply template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt:17]: Is there a chance we're missing some ktor config then?

Comment (Filmaluco) [reply template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt:17]: going to check

File: template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt (hunk 12 above)
Comment (Filmaluco) [reply template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt:76]: 1st is a list of callbacks for additional loggers other than the setup up one. (Like trees in Timber)
2nd to all log handlers that exist, like the one we creat in the `.setup()`

File: template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt (hunk 14 above)
Comment (Filmaluco) [reply template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt:56]: "FINE is a message level providing tracing information."  not the default...

File: template/core/domain/src/main/java/io/bloco/core/domain/models/Book.kt (hunk 27 above)
Comment (LilianaFaustinoBloco) [reply template/core/domain/src/main/java/io/bloco/core/domain/models/Book.kt:15]: This key is the ID that we use to request book details and it comes with this prefix from API.

I decided to remove it for consistency, so as not to be saving a useless piece of the string. Still, we might not remove and just change the book detail request path (that is `/works/$id.json`).

File: template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt (hunk 12 above)
Comment (sdsantos) [reply template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt:76]: Why have 2 different types of loggers?

Comment (Filmaluco) [reply template/core/commons/src/main/java/io/bloco/core/commons/Logger.kt:76]: if we want to log files to a file and store it for instance.
I can add as an example for datastore :D

File: template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt (hunk 24 above)
Comment (Filmaluco) [reply template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt:24]: I think this Exception is to cover the cast of the requested result body()... 
so I guess we can have it more specific by validation the status code like:

```
when (response.status) {
            HttpStatusCode.BadRequest -> {} // Throw errors or transform to T 
            HttpStatusCode.Conflict -> {}
            else -> throw this
        }
```

and/or only attempt to catch the cast exception.

but @LilianaFaustinoBloco developed this part, can you confirm me this?

Comment (LilianaFaustinoBloco) [reply template/core/data/src/main/java/io/bloco/core/data/network/OpenLibraryService.kt:24]: Yep, we can do this. 

But to handle this type of exception, I think we need a [HttpResponseValidator](https://ktor.io/docs/response-validation.html#non-2xx).
